reports/namespace-usage/main.go

```diff
@@ -0,0 +1,231 @@
+	if err != nil {
+		log.Fatalln("error in getting all namespaces from cluster", err.Error())
+	}
+
+	// Get the list of pods from the cluster which is set in the kclientset
+	podsList, err := namespace.GetAllPodsFromCluster(kclientset)
+	if err != nil {
+		log.Fatalln("error in getting all pods from cluster", err.Error())
+	}
+
+	nsReqMap := make(map[string]NamespaceResource, 0)
+
+	containerMap := make(map[string]int, 0)
+
+	// get resource request of each pod and container count
+	// and store it in namespaceResource map
+	for _, pod := range podsList {
+		req, namespace, newCount := GetPodResourceDetails(pod)
+		list := nsReqMap[namespace]
+		if _, exist := nsReqMap[namespace]; exist {
+			list.addNamespaceResource(req)
+			nsReqMap[namespace] = list
+		} else {
+			nsReqMap[namespace] = req
+		}
+		containerMap[namespace] += newCount
+	}
+
+	// Get top pods(resource used) of all namespaces from the cluster which is set in the mclientset
+	podMetricsList, err := namespace.GetAllPodMetricsesFromCluster(mclientset)
+	if err != nil {
+		log.Fatalln("error in getting all pod metrics from cluster", err.Error())
+	}
+
+	nsUsedMap := make(map[string]NamespaceResource, 0)
+
+	for _, podMetrics := range podMetricsList {
+		used, namespace := GetPodUsageDetails(podMetrics)
+		list := nsUsedMap[namespace]
+		if _, exist := nsUsedMap[namespace]; exist {
+			list.addNamespaceResource(used)
+			nsUsedMap[namespace] = list
+		} else {
+			nsUsedMap[namespace] = used
+		}
+	}
+
+	// get namespace quota of namespaces to find hard limits of pods from the cluster
+	rsQuotasList, err := namespace.GetAllResourceQuotasFromCluster(kclientset)
+	if err != nil {
+		log.Fatalln("error in getting all resourcequota from cluster", err.Error())
+	}
+
+	nsQuotaMap := make(map[string]NamespaceResource, 0)
+
+	for _, rsQuota := range rsQuotasList {
+		hardLimits, namespace, err := GetPodHardLimits(rsQuota)
+		if err != nil {
+			log.Fatalln(err.Error())
+		}
+		nsQuotaMap[namespace] = hardLimits
+	}
+
+	var usageReports []UsageReport
+	// Build the total usageReport
+	for _, ns := range nsList {
+		var usageReport UsageReport
+		usageReport.Name = ns.Name
+		usageReport.Requested = nsReqMap[ns.Name]
+		usageReport.Used = nsUsedMap[ns.Name]
+		usageReport.Hardlimits = nsQuotaMap[ns.Name]
+		usageReport.ContainerCount = containerMap[ns.Name]
+		usageReports = append(usageReports, usageReport)
+	}
+
+	jsonToPost, err := BuildJsonMap(usageReports)
+	if err != nil {
+		log.Fatalln(err.Error())
+	}
+
+	// Post json to hoowdaw api
+	err = hoodaw.PostToApi(jsonToPost, hoodawApiKey, &endPoint)
+	if err != nil {
+		log.Fatalln(err.Error())
+	}
+}
+
+// GetPodResourceDetails takes a Pod of type v1.Pod and collect
+// all resources summed up for all containers of the pod and return the result
+func GetPodResourceDetails(pod v1.Pod) (r NamespaceResource, namespace string, containerCount int) {
+	reqs, _ := v1.ResourceList{}, v1.ResourceList{}
```

Comment:
This is interesting, what does this do?

Reply:
This one parse the pod of type v1.pod, loop though all containers, get the resources requests(cpu,memory), add them up and return.
The `addResourceList` is a copied and modified function from kubectl package.

Reply:
the `reqs, _ := ...` will assign one variable and drop the other?

Reply:
I  realised that just a typo. Removed that.